### ISSUE TITLE
Change the process for merging the upstream

### ISSUE THREAD

**Issue by piersy** [autonity/autonity#477]
# Problem

TL;DR
1. Future incorporation of changes from upstream will be laborious. (try running `git merge upstream/master` while on `develop` and see how many conflicts you get)
2. We don't have a way to see what code has been worked on by the autonity team as opposed to the geth team.

## Detailed description

Our merging process is not maintaining proper merge commits (commits that have 2 parent commits), instead we seem to have been squashing or re-basing our merge commits such that they lose the parent commit of the remote upstream and the merge tag.

When you run `git merge X` git looks for the most recent commit shared by your currently checked out branch and `X` this is called the merge-base. Git then applies all the changes from the merge-base to the tip of `X` to your branch and inserts a merge commit that joins the 2 histories.

We can see how this works with with a diagram, `develop` (the branch on the left) is merged into `429` twice, each time it is merged git looks back to the commit where the 2 branches were previously joined and only merges changes from that point onward. This is the intended way to manage merges.
 
```git
| * 579be743a Added truffle tests to test user status up/downgrading. Tests all ...
| * 4d90cf813 The 'test validator can get users list' truffle test incorrectly t...
| *   2be637595 Merge branch 'develop' into 429
| |\  
| |/  
|/|   
* | da9970f09 change behaviour of codecov (#440)
* | 1114f3f23 Restore linters (#446)
* | a9ebc6620 Ignore miner leak (#443)
| *   aca49e36f Merge branch 'develop' into 429
| |\  
| |/  
|/|   
* | c545125a6 (tag: dev-0.3.2) Fix Codecov reports (#425)
* | ec8eb713a Fix codecoverage and integrate with Github actions (#417)
* | 9babfb718 Remove open network flag (#411)
| * 2394b32b3 Change the changeUserStatus function to simply remove the user and...
| * 840fdb7aa Modify the createUser function to check for duplicates before addi...
|/  
* fdc7c36d0 (HEAD -> master, origin/voting-power-calculation, origin/master) Fix...
```
Our merges from the upstream however only have one parent. They appear as just normal (although very large) commits on our branch.

Git cannot detect these commits as merge commits and goes to the most recent merge-base to determine the changes to merge.

```
$ git merge-base develop upstream/master
24d727b6d6e2c0cde222fa12155c4a6db5caaf2e
```

That commit was made in Dec 2018, so running a merge will try to apply all changes from geth `master` from Dec 2018 to now. The last time we pulled in changes from geth without creating a proper merge commit was Jan 23 2020. Git perceives all changes in our `develop` as ours and all changes form geth as theirs, this causes potential conflicts on all changes from Dec 2018 to Jan 2020.

Because there is no record in our history of what commit we pulled in changes from geth, we also cannot see what files are files exclusively worked on by geth  vs autonity.  Diffing 1.9 release from geth with `develop`, removing autonity only files and adjusting for import paths gives 328 changed go files out of a possible 895 go files in the geth codebase.

# Fix

The fix is to merge again and preserve the merge commit. This will mean that the next merge will use that as the merge base, and we will be able to diff `develop` with the merge base to see what is the work done by the autonity team.

I believe git blame will still attribute all changes that we improperly merged to the committer. I can't see any way to fix this (suggestions welcome)

# Acceptance Criteria

Merge geth upstream into develop and preserve the merge commit (no rebasing or squashing)

**Comment by piersy** [comment]
I think I found the last commit merged from go-ethereum master branch.

Its `3bb6815fc` I created a branch `compare_branch` where I reverted all the changed import paths.

So running `git diff 3bb6815fc origin/compare_branch` will show you what differs between geth and autonity.

**Comment by piersy** [comment]
So I think I found a good solution to our problem. It's a feature of git that I was not aware of called grafts and it allows you to override the merge base when doing a merge.

This means we can manually specify the merge base to the last merged commit from `go-ethereum/master` and then perfrorm a merge, which will behave as if we had maintained the original merge commits.

I've just tried it and it seemed to work, I only got 59 conflicts without any `both deleted` or `both added` conflicts.

The resultant commit is a merge commit, meaning that the next time we come to merge we can just run `merge upstream/master` and only deal with the relevant changes.

An example of using grafts can be seen [here](https://stackoverflow.com/a/32359327/370364)
